dm/master/scheduler/scheduler.go

```diff
@@ -0,0 +1,1139 @@
+	// - when bounded the source to a worker.
+	unbounds map[string]struct{}
+
+	// expectant relay stages for sources, source ID -> stage.
+	// add:
+	// - bound the source to a worker (at first time).
+	// - recover from etcd (calling `recoverSources`).
+	// update:
+	// - update stage by user request (calling `UpdateExpectRelayStage`).
+	// delete:
+	// - remove source by user request (calling `RemoveSourceCfg`).
+	expectRelayStages map[string]ha.Stage
+
+	// expectant subtask stages for tasks & sources, task name -> source ID -> stage.
+	// add:
+	// - add/start subtask by user request (calling `AddSubTasks`).
+	// - recover from etcd (calling `recoverSubTasks`).
+	// update:
+	// - update stage by user request (calling `UpdateExpectSubTaskStage`).
+	// delete:
+	// - remove/stop subtask by user request (calling `RemoveSubTasks`).
+	expectSubTaskStages map[string]map[string]ha.Stage
+}
+
+// NewScheduler creates a new scheduler instance.
+func NewScheduler(pLogger *log.Logger) *Scheduler {
+	return &Scheduler{
+		logger:              pLogger.WithFields(zap.String("component", "scheduler")),
+		sourceCfgs:          make(map[string]config.MysqlConfig),
+		subTaskCfgs:         make(map[string]map[string]config.SubTaskConfig),
+		workers:             make(map[string]*Worker),
+		bounds:              make(map[string]*Worker),
+		unbounds:            make(map[string]struct{}),
+		expectRelayStages:   make(map[string]ha.Stage),
+		expectSubTaskStages: make(map[string]map[string]ha.Stage),
+	}
+}
+
+// Start starts the scheduler for work.
+func (s *Scheduler) Start(pCtx context.Context, etcdCli *clientv3.Client) error {
+	s.logger.Info("the scheduler is starting")
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.started {
+		return terror.ErrSchedulerStarted.Generate()
+	}
+
+	s.reset() // reset previous status.
+
+	// recover previous status from etcd.
+	err := s.recoverSources(etcdCli)
+	if err != nil {
+		return err
+	}
+	err = s.recoverSubTasks(etcdCli)
+	if err != nil {
+		return err
+	}
+	rev, err := s.recoverWorkersBounds(etcdCli)
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithCancel(pCtx)
+
+	// starting to observe status of DM-worker instances.
+	workerEvCh := make(chan ha.WorkerEvent, 10)
+	workerErrCh := make(chan error, 10)
+	s.wg.Add(2)
+	go func() {
+		defer func() {
+			s.wg.Done()
+			close(workerEvCh)
+			close(workerErrCh)
+		}()
+		ha.WatchWorkerEvent(ctx, etcdCli, rev+1, workerEvCh, workerErrCh)
+	}()
+	go func() {
+		defer s.wg.Done()
+		s.handleWorkerEv(ctx, workerEvCh, workerErrCh)
+	}()
+
+	s.started = true // started now
+	s.cancel = cancel
+	s.etcdCli = etcdCli
+	s.logger.Info("the scheduler has started")
+	return nil
+}
+
+// Close closes the scheduler.
+func (s *Scheduler) Close() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if !s.started {
+		return
+	}
+
+	s.logger.Info("the scheduler is closing")
+	if s.cancel != nil {
+		s.cancel()
+		s.cancel = nil
+	}
+
+	s.wg.Wait()
```

Comment:
some function needs lock in `handleWorkerEv`, this may cause dead lock.

Reply:
this `s.wg.Wait()` should wait `handleWorkerEv` return.
```go
go func() {
	defer s.wg.Done()
	s.handleWorkerEv(ctx, workerEvCh, workerErrCh)
}()
```

Reply:
for example:
1. Close  -> lock
2. execute `handleWorkerOffline in` `handleWorkerEv` -> lock, this will wait lock
3. s.wg.wait, wait `handleWorkerEv` finish, this will cause dead lock

Reply:
Good catchup, fixed in https://github.com/pingcap/dm/pull/473/commits/bdbe56f3e6640aaa9d066ab48cd666a81c8399a1.